### PR TITLE
Restructure training scripts to avoid duplication

### PR DESCRIPTION
**Pull request by ianjjohnson**
This makes training scripts no-longer model-specific, but instead configurable per-model.
Users can still contribute improvements to scripts via:
- New scripts
- Improvements to existing scripts
- New parameters which improve on script performance

The only major outstanding question I have is: what should we do with historical training metadata when a script is improved? Perhaps the metadata for each run should include a script version based on a commit hash?

Let me know your thoughts.

(This also makes a few tweaks to basic_training to make it modular and switch to a DecreaseLROnPlateau strategy for learning rate).

@LukeWood

### CODE REVIEW

File: examples/training/classification/basic_imagenet_training.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+    "script_authors": ["ianjjohnson"]
+    "densenet121": {
```
Comment (LukeWood) [inline examples/training/classification/basic_imagenet_training.json:3]: so every model will land here basically?

File: examples/training/classification/basic_imagenet_training.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+    "script_authors": ["ianjjohnson"]
```
Comment (LukeWood) [inline examples/training/classification/basic_imagenet_training.json:2]: I do like the idea of "authors" to attribute credit to the entire group that wrote it.

File: examples/training/classification/basic_imagenet_training.py [hunk 3]
```diff
@@ -161,7 +172,8 @@ def augment(img, label):
 strategy = tf.distribute.MirroredStrategy()
 
 with strategy.scope():
-    model = DenseNet121(
+    model = eval(f"models.{FLAGS.model_name}")
```
Comment (LukeWood) [inline examples/training/classification/basic_imagenet_training.py:175]: can we use __dict__ here too?

File: examples/training/classification/basic_imagenet_training.py [hunk 4]
```diff
@@ -196,7 +203,7 @@ def augment(img, label):
 """
 
 
-metrics = [metrics.CategoricalAccuracy()]
+metrics = ["accuracy"]
```
Comment (LukeWood) [inline examples/training/classification/basic_imagenet_training.py:206]: I think Francois wanted to use a specific metric instance right?  lets keep it in line with the other ones

File: examples/training/classification/basic_imagenet_training.py [hunk 5]
```diff
@@ -206,6 +213,9 @@ def augment(img, label):
 
 
 callbacks = [
+    callbacks.ReduceLROnPlateau(
```
Comment (LukeWood) [inline examples/training/classification/basic_imagenet_training.py:216]: patience=20 will generalize better.  Augmentation is very stochastic.

File: examples/training/classification/basic_imagenet_training.py [hunk 6]
```diff
@@ -206,6 +213,9 @@ def augment(img, label):
...
+        monitor="val_loss", factor=0.3, patience=5, min_lr=0.0001
+    ),
     callbacks.EarlyStopping(patience=30),
```
Comment (LukeWood) [inline examples/training/classification/basic_imagenet_training.py:219]: lets increase patience now that we use callbacks.ReduceLROnPlateau()

**Review by LukeWood — CHANGES_REQUESTED** [review]
Some comments, thanks Ian!

File: examples/training/classification/basic_imagenet_training.json (hunk 1 above)
Comment (ianjjohnson) [reply examples/training/classification/basic_imagenet_training.json:3]: Right -- every model trained using this script.

File: examples/training/classification/basic_imagenet_training.py (hunk 4 above)
Comment (ianjjohnson) [reply examples/training/classification/basic_imagenet_training.py:206]: Yes, but it turns out that this throws an error because if you do it with a real object you have to do it in the same strategy scope as you instantiate the model.

I'll just update this to re-open the scope when instantiating this.

File: examples/training/classification/basic_imagenet_training.py (hunk 6 above)
Comment (ianjjohnson) [reply examples/training/classification/basic_imagenet_training.py:219]: I've updated this to 40 (basically meaning two consecutive learning rate drops without improvement)

File: examples/training/classification/basic_imagenet_training.py (hunk 5 above)
Comment (ianjjohnson) [reply examples/training/classification/basic_imagenet_training.py:216]: Done.

File: examples/training/classification/basic_imagenet_training.py [hunk 7]
```diff
@@ -47,22 +47,32 @@
...
-    "backup_path", None, "Directory which will be used for training backups"
+    "backup_path", None, "Directory which will be used for training backups."
 )
 flags.DEFINE_string(
     "weights_path", None, "Directory which will be used to store weight checkpoints."
 )
 flags.DEFINE_string(
     "tensorboard_path", None, "Directory which will be used to store tensorboard logs."
 )
+flags.DEFINE_integer("batch_size", 256, "Batch size for training and evaluation.")
```
Comment (LukeWood) [inline examples/training/classification/basic_imagenet_training.py:63]: I'm not a huge fan of "basic_imagenet_training.py"

can we use the structured `examples/training/classification/imagenet/basic_training.py`, include a README providing an overview of the task, say we use .tfrecord loading, etc.

Then in subdirectories we can have specific scripts: "ResNet50"

Comment (LukeWood) [reply examples/training/classification/basic_imagenet_training.py:63]: basic_imagenet_training.py is a good name for a keras.io tutorial - but we can do that in keras.io when we move there.

Comment (ianjjohnson) [reply examples/training/classification/basic_imagenet_training.py]: Done -- restructured again as discussed

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
looks good to me!  Thanks Ian!
